SerializedScriptValue could be faster by atomizing strings early in some cases

cdumez opened the pull request:
#### d2a72f55c356ddf05a64fb2b0ca44ca3fb03875f
<pre>
SerializedScriptValue could be faster by atomizing strings early in some cases
<a href="https://bugs.webkit.org/show_bug.cgi?id=259267">https://bugs.webkit.org/show_bug.cgi?id=259267</a>
rdar://112759042

Reviewed by Darin Adler.

Upon deserialization, in the `ObjectStartVisitMember` case, we would call
`readStringData()` which would allocate a String from the raw characters
and we would later construct a JSC::Identifier from it using
JSC::Identifier::fromString(), which would then atomize the string.

To avoid unnecessary string allocations in the case where the string is
already in the AtomString table, we now add a `ShouldAtomize` parameter
to `readStringData()`. Then the parameter is `Yes`, it now constructs an
AtomString right away instead of allocating a String. As a result, we
avoid unnecessary String allocations and the construction of a
JSC::Identifier is cheap using an already atomized string.

* Source/WebCore/bindings/js/SerializedScriptValue.cpp:
(WebCore::CloneSerializer::CloneSerializer):
(WebCore::CloneDeserializer::deserializeString):
(WebCore::CloneDeserializer::readString):
(WebCore::CloneDeserializer::readNullableString):
(WebCore::CloneDeserializer::readStringData):
(WebCore::CloneDeserializer::deserialize):

Canonical link: <a href="https://commits.webkit.org/266503@main">https://commits.webkit.org/266503@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e4f350a689e05478db62a3bede47e91e92b01972

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/13994 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/14309 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/14644 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/15732 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/13281 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/14078 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/16818 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/14393 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/15951 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/14164 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/14755 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/11857 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/16440 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/12041 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/12618 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/19647 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/13117 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/12782 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/15990 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/13329 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/11192 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/12604 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3391 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/16938 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/13173 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/bindings/js/SerializedScriptValue.cpp:
```diff
@@ -2959,31 +2966,33 @@ class CloneDeserializer : CloneBase {
             characters[i] = c;
         }
+        if (shouldAtomize == ShouldAtomize::Yes)
+            AtomString { str };
```
I think you want to write `str = AtomString { str }` here. But also, if this is at all common, we can write a more optimized version of this for short strings that doesn’t involve allocating and then deallocating memory when it’s a duplicate atom.

cdumez replied:
As far as I know, both x86-64 and Apple Silicon use little-endian so I didn't spend time optimizing this code path, though we could fairly easily do so.

cdumez committed:
SerializedScriptValue could be faster by atomizing strings early in some cases
https://bugs.webkit.org/show_bug.cgi?id=259267
rdar://112759042

Reviewed by Darin Adler.

Upon deserialization, in the `ObjectStartVisitMember` case, we would call
`readStringData()` which would allocate a String from the raw characters
and we would later construct a JSC::Identifier from it using
JSC::Identifier::fromString(), which would then atomize the string.

To avoid unnecessary string allocations in the case where the string is
already in the AtomString table, we now add a `ShouldAtomize` parameter
to `readStringData()`. Then the parameter is `Yes`, it now constructs an
AtomString right away instead of allocating a String. As a result, we
avoid unnecessary String allocations and the construction of a
JSC::Identifier is cheap using an already atomized string.

* Source/WebCore/bindings/js/SerializedScriptValue.cpp:
(WebCore::CloneSerializer::CloneSerializer):
(WebCore::CloneDeserializer::deserializeString):
(WebCore::CloneDeserializer::readString):
(WebCore::CloneDeserializer::readNullableString):
(WebCore::CloneDeserializer::readStringData):
(WebCore::CloneDeserializer::deserialize):

Canonical link: https://commits.webkit.org/266503@main
